File: skellig-test-step-processing/src/test/java/org/skellig/teststep/processing/converter/TestStepStateValueConverterTest.java
package org.skellig.teststep.processing.converter;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.valueextractor.TestStepValueExtractor;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class TestStepStateValueConverterTest {

    private TestStepStateValueConverter testStepStateValueConverter;
    private TestScenarioState testScenarioState;
    private TestStepValueExtractor valueExtractor;

    @BeforeEach
    void setUp() {
        testScenarioState = mock(TestScenarioState.class);
        valueExtractor = mock(TestStepValueExtractor.class);
        testStepStateValueConverter = new TestStepStateValueConverter(testScenarioState, valueExtractor);
    }

    @Test
    void testGetSimpleValueFromState() {
        String expectedResult = "v1";
        when(testScenarioState.get("key")).thenReturn(Optional.of(expectedResult));

        assertEquals(expectedResult, testStepStateValueConverter.convert("get(key)"));
    }

    @Test
    void testGetObjectValueFromState() {
        Object expectedResult = new Object();
        when(testScenarioState.get("key")).thenReturn(Optional.of(expectedResult));

        assertEquals(expectedResult, testStepStateValueConverter.convert("get(key)"));
    }

    @Test
    void testGetValueFromStateWithAttachedString() {
        when(testScenarioState.get("key")).thenReturn(Optional.of(Collections.singletonList("_")));

        assertEquals("^[_]^", testStepStateValueConverter.convert("^get(key)^"));
    }

    @Test
    void testGetValueFromStateWithExtractorAndAttachedString() {
        List<String> value = Collections.singletonList("_");
        when(valueExtractor.extract(value, "([0])")).thenReturn(value.get(0));
        when(testScenarioState.get("key")).thenReturn(Optional.of(value));

        assertEquals("^_^", testStepStateValueConverter.convert("^get(key).([0])^"));
    }

    @Test
    void testGetValueFromStateWithExtractor() {
        String value = "value";
        when(valueExtractor.extract(value, "(length)")).thenReturn(value.length());
        when(testScenarioState.get("key")).thenReturn(Optional.of(value));

        assertEquals(value.length(), testStepStateValueConverter.convert("get(key).(length)"));
    }
}

File: skellig-test-step-processing/src/test/kotlin/org/skellig/teststep/processing/converter/TestStepStateValueConverterTest.kt
package org.skellig.teststep.processing.converter

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.skellig.teststep.processing.exception.TestDataConversionException
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.valueextractor.TestStepValueExtractor

class TestStepStateValueConverterTest {

    private var testStepStateValueConverter: TestStepStateValueConverter? = null
    private var testScenarioState: TestScenarioState? = null

    @BeforeEach
    fun setUp() {
        testScenarioState = mock()
        testStepStateValueConverter = TestStepStateValueConverter(testScenarioState!!)
    }

    @Test
    fun testGetSimpleValueFromStateWhenKeyNull() {
        val ex = Assertions.assertThrows(TestDataConversionException::class.java) {
            testStepStateValueConverter!!.execute("get", arrayOf(null))
        }
        Assertions.assertEquals("No data found in Test Scenario State with key `null`", ex.message)
    }

    @Test
    fun testGetSimpleValueFromState() {
        val expectedResult = "v1"
        whenever(testScenarioState!!.get("key")).thenReturn(expectedResult)

        Assertions.assertEquals(expectedResult, testStepStateValueConverter!!.execute("get", arrayOf("key")))
    }

    @Test
    fun testGetObjectValueFromState() {
        val expectedResult = Any()
        whenever(testScenarioState!!.get("key")).thenReturn(expectedResult)

        Assertions.assertEquals(expectedResult, testStepStateValueConverter!!.execute("get", arrayOf("key")))
    }

    @Test
    fun testGetValueFromStateWhenNotExist() {
        whenever(testScenarioState!!.get("key")).thenReturn(null)

        Assertions.assertThrows(TestDataConversionException::class.java) { testStepStateValueConverter!!.execute("get", arrayOf("key")) }
    }
}